business/authentication/token_auth_controller.go

```diff
@@ -0,0 +1,214 @@
+// unexpected errors.
+type AuthenticationFailureError struct {
+	// A description of the authentication failure
+	Reason string
+
+	// Wraps the error causing the authentication failure
+	Detail error
+}
+
+// Error returns the string representation of an AuthenticationFailureError
+func (e *AuthenticationFailureError) Error() string {
+	if e.Detail != nil {
+		return fmt.Sprintf("%s: %v", e.Reason, e.Detail)
+	}
+
+	return e.Reason
+}
+
+// NewTokenAuthController initializesa new controller for handling token authentication, with the
+// given persistor and the given businessInstantiator. The businessInstantiator can be nil and
+// the initialized contoller will use the business.Get function.
+func NewTokenAuthController(persistor SessionPersistor, businessInstantiator func(authInfo *api.AuthInfo) (*business.Layer, error)) *tokenAuthController {
+	if businessInstantiator == nil {
+		businessInstantiator = business.Get
+	}
+
+	return &tokenAuthController{
+		businessInstantiator: businessInstantiator,
+		SessionStore:         persistor,
+	}
+}
+
+// Authenticate handles an HTTP request that contains a token passed in the "token" field of form data of
+// the body of the request (POST, PATCH or PUT methods). The token should be valid to be used in the
+// Kubernetes API, thus the token is verified by trying a request to the Kubernetes API.
+// If the Kubernetes API rejects the token, authentication fails with an invalid/expired token error. If
+// the token is accepted, privileges to read some namespace is checked. If some namespace is readable,
+// authentication succeeds and a session is started; else, authentication is rejected because the
+// user won't be able to see any data in Kiali.
+// An AuthenticationFailureError is returned if the authentication request is rejected (unauthorized). Any
+// other kind of error means that something unexpected happened.
+func (c tokenAuthController) Authenticate(r *http.Request, w http.ResponseWriter) (*UserSessionData, error) {
+	// Get the token from HTTP form data
+	err := r.ParseForm()
+	if err != nil {
+		return nil, fmt.Errorf("error parsing form data from client: %w", err)
+	}
+
+	token := r.PostForm.Get("token")
+	if token == "" {
+		return nil, errors.New("token is empty")
+	}
+
+	// Create a bs layer with the received token to check its validity.
+	bs, err := c.businessInstantiator(&api.AuthInfo{Token: token})
+	if err != nil {
+		return nil, fmt.Errorf("error instantiating the business layer: %w", err)
+	}
+
+	// Using the namespaces API to check if token is valid. In Kubernetes, the version API seems to allow
+	// anonymous access, so it's not feasible to use the version API for token verification.
+	nsList, err := bs.Namespace.GetNamespaces()
+	if err != nil {
+		c.SessionStore.TerminateSession(r, w)
+		return nil, &AuthenticationFailureError{Reason: "token is not valid or is expired", Detail: err}
+	}
+
+	// If namespace list is empty, return authentication failure.
+	if len(nsList) == 0 {
```

Comment:
Does kube return an empty list if the user does not have permissions to list all namespaces? Otherwise I don't think this situation is possible given that if Kiali is deployed then there must be a namespace it is deployed in.

Reply:
Not kube directly. But yes, an empty list can be returned.

That `bs.Namespace.GetNamespaces()` is a Kiali's function. Open it and you will see an interesting code playing with tokens :smile: